// @ngrx/effects/esm5/src/effects_feature_module.ngfactory.js
/**
 * @fileoverview This file was generated by the Angular template compiler. Do not edit.
 *
 * @suppress {suspiciousCode,uselessCode,missingProperties,missingOverride,checkTypes}
 * tslint:disable
 */ 
export { Inject_7 as Inject_1, ɵangular_packages_core_core_a_8 as ɵangular_packages_core_core_a_2, BehaviorSubject_6 as BehaviorSubject_8, Observable_1 as Observable_10 } from "ngrx/modules/store/src/reducer_manager.ngfactory";
export { Optional as Optional_4 } from "@angular/core";
export { StoreRootModule as StoreRootModule_6, StoreFeatureModule as StoreFeatureModule_23 } from "@ngrx/store/src/store_module";
export { ActionsSubject as ActionsSubject_7 } from "@ngrx/store/src/actions_subject";
export { ReducerObservable as ReducerObservable_9, ReducerManager as ReducerManager_15, ReducerManagerDispatcher as ReducerManagerDispatcher_20 } from "@ngrx/store/src/reducer_manager";
export { ScannedActionsSubject as ScannedActionsSubject_11 } from "@ngrx/store/src/scanned_actions_subject";
export { Subject_1 as Subject_12 } from "ngrx/modules/store/src/scanned_actions_subject.ngfactory";
export { Store as Store_13 } from "@ngrx/store/src/store";
export { StateObservable as StateObservable_14 } from "@ngrx/store/src/state";
export { INITIAL_STATE as INITIAL_STATE_16, INITIAL_REDUCERS as INITIAL_REDUCERS_18, REDUCER_FACTORY as REDUCER_FACTORY_19, STORE_FEATURES as STORE_FEATURES_24, FEATURE_REDUCERS as FEATURE_REDUCERS_25 } from "@ngrx/store/src/tokens";
export { InjectionToken_1 as InjectionToken_17 } from "ngrx/modules/store/src/tokens.ngfactory";
import * as i0 from "@angular/core";
import * as i1 from "ngrx/modules/effects/src/effects_feature_module";
import * as i2 from "ngrx/modules/effects/src/effects_root_module";
import * as i3 from "ngrx/modules/effects/src/tokens";
import * as i4 from "@ngrx/store/src/store_module";
var EffectsFeatureModuleNgFactory = i0.ɵcmf(i1.EffectsFeatureModule, [], function (_l) { return i0.ɵmod([i0.ɵmpd(512, i0.ComponentFactoryResolver, i0.ɵCodegenComponentFactoryResolver, [[8, []], [3, i0.ComponentFactoryResolver], i0.NgModuleRef]), i0.ɵmpd(1073742336, i1.EffectsFeatureModule, i1.EffectsFeatureModule, [i2.EffectsRootModule, i3.FEATURE_EFFECTS, [2, i4.StoreRootModule], [2, i4.StoreFeatureModule]])]); });
export { EffectsFeatureModuleNgFactory as EffectsFeatureModuleNgFactory };
