// @ngrx/effects/esm2015/src/effects_root_module.ngfactory.js
/**
 * @fileoverview This file was generated by the Angular template compiler. Do not edit.
 *
 * @suppress {suspiciousCode,uselessCode,missingProperties,missingOverride,checkTypes}
 * tslint:disable
 */ 
export { Inject_7 as Inject_2, ɵangular_packages_core_core_a_8 as ɵangular_packages_core_core_a_3, Observable_1 as Observable_9, BehaviorSubject_6 as BehaviorSubject_12 } from "ngrx/modules/store/src/reducer_manager.ngfactory";
export { Optional as Optional_5 } from "@angular/core";
export { Store as Store_8 } from "@ngrx/store/src/store";
export { StateObservable as StateObservable_10 } from "@ngrx/store/src/state";
export { ActionsSubject as ActionsSubject_11 } from "@ngrx/store/src/actions_subject";
export { ReducerManager as ReducerManager_13, ReducerManagerDispatcher as ReducerManagerDispatcher_18, ReducerObservable as ReducerObservable_22 } from "@ngrx/store/src/reducer_manager";
export { INITIAL_STATE as INITIAL_STATE_14, INITIAL_REDUCERS as INITIAL_REDUCERS_16, REDUCER_FACTORY as REDUCER_FACTORY_17, STORE_FEATURES as STORE_FEATURES_26, FEATURE_REDUCERS as FEATURE_REDUCERS_27 } from "@ngrx/store/src/tokens";
export { InjectionToken_1 as InjectionToken_15 } from "ngrx/modules/store/src/tokens.ngfactory";
export { StoreRootModule as StoreRootModule_21, StoreFeatureModule as StoreFeatureModule_25 } from "@ngrx/store/src/store_module";
export { ScannedActionsSubject as ScannedActionsSubject_23 } from "@ngrx/store/src/scanned_actions_subject";
export { Subject_1 as Subject_24 } from "ngrx/modules/store/src/scanned_actions_subject.ngfactory";
import * as i0 from "@angular/core";
import * as i1 from "ngrx/modules/effects/src/effects_root_module";
import * as i2 from "ngrx/modules/effects/src/effect_sources";
import * as i3 from "ngrx/modules/effects/src/effects_runner";
import * as i4 from "@ngrx/store/src/store";
import * as i5 from "ngrx/modules/effects/src/tokens";
import * as i6 from "@ngrx/store/src/store_module";
var EffectsRootModuleNgFactory = i0.ɵcmf(i1.EffectsRootModule, [], function (_l) { return i0.ɵmod([i0.ɵmpd(512, i0.ComponentFactoryResolver, i0.ɵCodegenComponentFactoryResolver, [[8, []], [3, i0.ComponentFactoryResolver], i0.NgModuleRef]), i0.ɵmpd(1073742336, i1.EffectsRootModule, i1.EffectsRootModule, [i2.EffectSources, i3.EffectsRunner, i4.Store, i5.ROOT_EFFECTS, [2, i6.StoreRootModule], [2, i6.StoreFeatureModule]])]); });
export { EffectsRootModuleNgFactory as EffectsRootModuleNgFactory };
